packages/react/src/search/useSearch.ts

```diff
@@ -0,0 +1,196 @@
+  isError: boolean;
+  /**
+   * Total number of results across all pages.
+   */
+  total: number;
+  /**
+   * Total number of pages available based on `total` and `pageSize`.
+   */
+  totalPages: number;
+  /**
+   * The error object if the last search request failed, or null if no error occurred.
+   */
+  error: Error | null;
+}
+
+/**
+ * React hook for performing search queries with pagination.
+ * @param {UseSearchOptions} options - Configuration options for the search hook.
+ * @returns {UseSearchState} The search state.
+ * @public
+ */
+export const useSearch = <T = GenericFields>(
+  options: UseSearchOptions<T>
+): UseSearchState<T> => {
+  const { query, page = DEFAULT_PAGE, searchIndexId, pageSize = DEFAULT_PAGE_SIZE, sort } = options;
+
+  if (!searchIndexId) {
+    throw new Error('useSearch: searchIndexId is required');
+  }
+
+  const [state, setState] = useState<{
+    results: T[];
+    isLoading: boolean;
+    isSuccess: boolean;
+    isError: boolean;
+    total: number;
+    totalPages: number;
+    error: Error | null;
+  }>({
+    results: [],
+    isLoading: true,
+    isSuccess: false,
+    isError: false,
+    total: 0,
+    totalPages: 0,
+    error: null,
+  });
+
+  const searchService = useSearchService();
+  const abortControllerRef = useRef<AbortController | null>(null);
+
+  const search = useCallback(async () => {
+    if (abortControllerRef.current) {
+      abortControllerRef.current.abort();
+    }
+
+    abortControllerRef.current = new AbortController();
+    const signal = abortControllerRef.current.signal;
+
+    setState({
+      results: [],
+      isLoading: true,
+      isSuccess: false,
+      isError: false,
+      total: 0,
+      totalPages: 0,
+      error: null,
+    });
+
+    try {
+      const offset = getOffset(page, pageSize);
+
+      const searchParams: SearchParameters<T> = {
+        searchIndexId,
+        keyphrase: query,
+        limit: pageSize,
+        offset,
+        sort,
+      };
+
+      const { results: searchResults, total } = await searchService.search<T>(searchParams, {
+        signal,
+      });
+
+      if (signal.aborted) {
+        return;
+      }
+
+      const totalPages = Math.ceil(total / pageSize);
+
+      setState({
+        results: searchResults,
+        total,
+        totalPages,
+        isSuccess: true,
+        isError: false,
+        isLoading: false,
+        error: null,
+      });
+    } catch (err) {
+      // Don't set error if request was aborted
+      if (signal.aborted) {
+        return;
+      }
+
+      const errorMessage = err instanceof Error ? err : new Error('Search failed');
+      setState({
+        results: [],
+        total: 0,
+        totalPages: 0,
+        isSuccess: false,
+        isError: true,
+        isLoading: false,
+        error: errorMessage,
+      });
+    }
+  }, [searchService, searchIndexId, page, pageSize, sort, query]);
+
+  useEffect(() => {
+    search();
+  }, [search]);
+
+  useEffect(() => {
```

Comment:
what is the point of separating unsubscribe logic into separate effect? It can lead to a faulty behavior as abort controller will be triggered only on component unmount, but effect with search will be executed each time search parameter is changed. So current abort controller doesn't prevent race conditions and "stale state setting".
Move abortion to the main `useEffect`. You also can make search function to accept abort controller - this way you can get rid of `useRef`